contracts/UniswapV3Staker.sol

```diff
@@ -218,43 +218,85 @@ contract UniswapV3Staker is
             'position is invalid'
         );
 
+        (bytes32 incentiveId, Incentive memory incentive, Stake memory stake) =
+            _getUpdateRewardParams(params);
+        (uint128 reward, uint160 secondsInPeriodX128, ) =
+            getRewardInfo(incentive, stake, params);
+
+        _updateIncentive(incentiveId, secondsInPeriodX128, reward, incentive.totalRewardUnclaimed);
+
+        delete stakes[params.tokenId][incentiveId];
         deposits[params.tokenId].numberOfStakes -= 1;
+        rewards[incentive.rewardToken][msg.sender] = uint128(
+            SafeMath.add(rewards[incentive.rewardToken][msg.sender], reward)
+        );
+        emit TokenUnstaked(params.tokenId);
+    }
 
-        (address poolAddress, int24 tickLower, int24 tickUpper, ) =
-            _getPositionDetails(params.tokenId);
+    function claimRewardFromExistingStake(
+        StakeParams memory params,
+        address to
+    ) external {
+        require(
+            deposits[params.tokenId].owner == msg.sender,
+            'NOT_YOUR_DEPOSIT'
+        );
+        (bytes32 incentiveId, Incentive memory incentive, Stake memory stake) =
+            _getUpdateRewardParams(params);
+        (
+            uint128 reward,
+            uint160 secondsInPeriodX128,
+            uint160 secondsPerLiquidityInsideX128
+        ) = getRewardInfo(incentive, stake, params);
+
+        _updateIncentive(incentiveId, secondsInPeriodX128, reward, incentive.totalRewardUnclaimed);
+
+        stakes[params.tokenId][incentiveId].secondsPerLiquidityInitialX128 =
+            secondsPerLiquidityInsideX128 +
+            1;
+        TransferHelper.safeTransfer(incentive.rewardToken, to, reward);
+        emit RewardClaimedFromExistingStake(to, reward);
```

Comment:
Is it necessary to have both `RewardClaimed` and `RewardClaimedFromExistingStake` ?